Sources/MessagingInApp/State/InAppMessageMiddleware.swift

```diff
@@ -0,0 +1,232 @@
+import CioInternalCommon
+import Foundation
+
+func userAuthenticationMiddleware() -> InAppMessageMiddleware {
+    middleware { _, getState, next, action in
+        let state = getState()
+
+        // Block actions that require valid userId unless userId is set.
+        switch action {
+        case .initialize,
+             .setUserIdentifier,
+             .setPageRoute,
+             .resetState:
+            return next(action)
+
+        default:
+            let userId = state.userId
+            guard let userId = userId, !userId.isBlankOrEmpty() else {
+                return next(.reportError(message: "[InApp] Blocked action: \(action) because userId (\(String(describing: userId))) is invalid"))
+            }
+
+            return next(action)
+        }
+    }
+}
+
+func routeMatchingMiddleware(logger: CioInternalCommon.Logger) -> InAppMessageMiddleware {
+    middleware { dispatch, getState, next, action in
+        let state = getState()
+        // Check for page rule match if the action is setting new route and userId is set.
+        guard case .setPageRoute = action, let userId = state.userId, !userId.isBlankOrEmpty() else {
+            return next(action)
+        }
+
+        // Update current route first
+        next(action)
+
+        // Check if there is a message currently displayed and if it has a page rule
+        let currentRoute = state.currentRoute
+        let currentMessage: Message? = state.currentMessageState.activeModalMessage
+        let doesCurrentMessageRouteMatch: Bool = currentMessage.flatMap { message in
+            message.doesHavePageRule() && (currentRoute.flatMap(message.doesPageRuleMatch) ?? true)
+        } ?? true
```

Comment:
```suggestion
        
let doesCurrentMessageRouteMatch: Bool = {
    guard let message = currentMessage else {
        return true  // No current message, so we consider it a match
    }
    
    guard message.doesHavePageRule() else {
        return true  // Message has no page rule, so we consider it a match
    }
    
    guard let route = currentRoute else {
        return false  // Message has a page rule, but there's no current route
    }
    
    return message.doesPageRuleMatch(route: route)
}()
```

One flatmap might have made it simpler but too just makes its hard to grasp, I tried to simplify and I feel like it makes a lot more sense, let me know .

Reply:
I think this can be further simplified. I revisited the code, along with Android implementation, and made it much simpler. I've also added comments to explain things further. Let me know if this looks better now.

Reply:
Looks great now